helix-term/src/commands.rs

```diff
@@ -2092,6 +2093,149 @@ fn symbol_picker(cx: &mut Context) {
+
+    cx.callback(
+        future,
+        move |editor: &mut Editor,
+              compositor: &mut Compositor,
+              response: Option<Vec<lsp::SymbolInformation>>| {
+            if let Some(symbols) = response {
+                let picker = Picker::new(
+                    symbols,
+                    |symbol| (&symbol.name).into(),
+                    move |editor: &mut Editor, symbol, _action| {
+                        push_jump(editor);
+                        let (view, doc) = current!(editor);
+
+                        // if let Some(range) =
+                        //     lsp_range_to_range(doc.text(), symbol.location.range, offset_encoding)
+                        // {
+                        //     doc.set_selection(view.id, Selection::single(range.to(), range.from()));
+                        //     align_view(doc, view, Align::Center);
+                        // }
+                    },
+                );
+                compositor.push(Box::new(picker))
+            }
+        },
+    )
+}
+
+pub fn code_action(cx: &mut Context) {
+    let (view, doc) = current!(cx.editor);
+
+    let language_server = match doc.language_server() {
+        Some(language_server) => language_server,
+        None => return,
+    };
+
+    let range = range_to_lsp_range(
+        doc.text(),
+        doc.selection(view.id).primary(),
+        language_server.offset_encoding(),
+    );
+
+    let future = language_server.code_actions(doc.identifier(), range);
+    let offset_encoding = language_server.offset_encoding().clone();
+
+    cx.callback(
+        future,
+        move |_editor: &mut Editor,
+              compositor: &mut Compositor,
+              response: Option<lsp::CodeActionResponse>| {
+            if let Some(actions) = response {
+                let picker = Picker::new(
+                    actions,
+                    |action| match action {
+                        lsp::CodeActionOrCommand::CodeAction(action) => {
+                            action.title.as_str().into()
+                        }
+                        lsp::CodeActionOrCommand::Command(command) => command.title.as_str().into(),
+                    },
+                    move |editor, code_action, _action| {
+                        make_code_action_callback(editor, code_action, offset_encoding)
+                    },
+                );
+                compositor.push(Box::new(picker))
+            }
+        },
+    )
+}
+
+fn make_code_action_callback(
+    editor: &mut Editor,
+    code_action: &lsp::CodeActionOrCommand,
+    offset_encoding: OffsetEncoding,
+) {
+    match code_action {
+        lsp::CodeActionOrCommand::Command(command) => {
+            todo!("command: {:?}", command);
+        }
+        lsp::CodeActionOrCommand::CodeAction(code_action) => {
+            log::debug!("code action: {:?}", code_action);
+            if let Some(ref edit) = code_action.edit {
```

Comment:
I think this could be rewritten with like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=39b88b0f6b48df391e5f69aed845319e

Reply:
or you can also pattern match 
```rust
if let (Some(lsp::CodeAction { edit: Edit { document_changes: Some(changes) )... 
```

Reply:
@archseer 
We need to use `document_changes` and `changes` independantly, so that wouldn't really work.

This function also changed a bit, I think it's not relevant anymore.